cli/src/spend_utils.rs

```diff
@@ -166,7 +168,25 @@ where
 {
     let (fee, compute_unit_info) = match blockhash {
         Some(blockhash) => {
-            let mut dummy_message = build_message(0);
+            // If the from account is the same as the fee payer, it's impossible
+            // to give a correct amount for the simulation with `SpendAmount::All`
+            // or `SpendAmount::RentExempt`.
+            // To know how much to transfer, we need to know the transaction fee,
+            // but the transaction fee is dependent on the amount of compute
+            // units used, which requires simulation.
+            // To get around this limitation, we simulate against an amount of
+            // `0`, since there are few situations in which `SpendAmount` can
+            // be `All` or `RentExempt` *and also* the from account is the fee
+            // payer.
+            let lamports = match amount {
+                SpendAmount::Some(lamports) => lamports,
+                SpendAmount::AllWithMinimum(lamports) => lamports,
+                SpendAmount::All | SpendAmount::RentExempt if from_pubkey == fee_pubkey => 0,
+                SpendAmount::All => from_balance,
+                SpendAmount::RentExempt => from_balance.saturating_sub(from_rent_exempt_minimum),
+            };
```

Comment:
I feel like this code would be easier to read if the `if from_pubkey == fee_pubkey` if statement wrapped the `match amount`, what do you think? Also I think that `AllWithMinimum` should probably just use `from_balance` when from != fee payer. Does that make sense?

```suggestion
            let lamports = if from_pubkey == fee_pubkey {
                match amount {
                    SpendAmount::Some(lamports) => lamports,
                    SpendAmount::AllWithMinimum(minimum_balance) => minimum_balance,
                    SpendAmount::All | SpendAmount::RentExempt => 0,
                }
            } else {
                match amount {
                    SpendAmount::Some(lamports) => lamports,
                    SpendAmount::AllWithMinimum(_) | SpendAmount::All => from_balance,
                    SpendAmount::RentExempt => {
                        from_balance.saturating_sub(from_rent_exempt_minimum)
                    }
                }
            };
```

Reply:
This is clearer, thanks for the suggestion